### PR TITLE
Adding custom tpu send transaction example

### PR DESCRIPTION
**Pull request by godmodegalactus**

### CODE REVIEW

File: services/src/transaction_service.rs [hunk 1]
```diff
@@ -122,6 +122,15 @@ pub struct TransactionService {
 
 impl TransactionService {
     pub async fn send_transaction(
+        &self,
+        tx: Transaction,
+        max_retries: Option<u16>,
```
Comment (grooviegermanikus) [inline services/src/transaction_service.rs:127]: nit: we should stick to one data type for this (i.e. `usize`)

File: services/src/tpu_utils/tpu_service.rs [hunk 2]
```diff
@@ -127,7 +128,7 @@ impl TpuService {
             .get_slot_leaders(current_slot, last_slot)
             .await?;
         // get next leader with its tpu port
-        let connections_to_keep = next_leaders
+        let connections_to_keep: HashMap<_, _> = next_leaders
```
Comment (grooviegermanikus) [inline services/src/tpu_utils/tpu_service.rs:131]: why do you need that?
maybe you could specify the full type in that case

File: cluster-endpoints/src/rpc_polling/vote_accounts_and_cluster_info_polling.rs [hunk 3]
```diff
@@ -11,10 +11,10 @@ pub fn poll_cluster_info(
 ) -> AnyhowJoinHandle {
     // task MUST not terminate but might be aborted from outside
     tokio::spawn(async move {
+        tokio::time::sleep(Duration::from_secs(1)).await;
```
Comment (grooviegermanikus) [inline cluster-endpoints/src/rpc_polling/vote_accounts_and_cluster_info_polling.rs:14]: pls add a comment why you need that

File: cluster-endpoints/src/rpc_polling/vote_accounts_and_cluster_info_polling.rs [hunk 4]
```diff
@@ -23,7 +23,7 @@ pub fn poll_cluster_info(
                 Err(error) => {
                     warn!("rpc_cluster_info failed <{:?}> - retrying", error);
                     // throttle
-                    tokio::time::sleep(Duration::from_secs(2500)).await;
+                    tokio::time::sleep(Duration::from_secs(10)).await;
```
Comment (grooviegermanikus) [inline cluster-endpoints/src/rpc_polling/vote_accounts_and_cluster_info_polling.rs:26]: would put that on a bugfix branch

File: core/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                }
+            }
+            Err(_elapsed) => {
+                debug!(
+                    "waiting for latest block info ({}) ... {:.02}ms",
+                    commitment_config.commitment,
+                    started.elapsed().as_secs_f32() * 1000.0
+                );
+            }
+            Ok(Err(_error)) => {
```
Comment (grooviegermanikus) [inline core/src/utils.rs:28]: **!!** you should handle Lagged and Closed  are differently:
* lagged: continue
* Closed: sleep some 10 second, then panic (create_grpc_multiplex_blocks_subscription _reconnect_loop_ never terminates unless cluster_endpoint_tasks get aborted which happens on shutdown; this method could gracefully wait if it detects shutdown)

File: core/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    loop {
+        match timeout(Duration::from_millis(500), blockinfo_stream.recv()).await {
+            Ok(Ok(block_info)) => {
+                if block_info.commitment_config == commitment_config {
+                    return block_info;
+                }
+            }
+            Err(_elapsed) => {
+                debug!(
+                    "waiting for latest block info ({}) ... {:.02}ms",
```
Comment (grooviegermanikus) [inline core/src/utils.rs:23]: pls add to log line what it's doing -> i.e. "timed out waiting for latest block info (...) - continue waiting"

File: core/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,33 @@
+use std::time::Duration;
+
+use log::debug;
+use solana_sdk::commitment_config::CommitmentConfig;
+use tokio::time::{timeout, Instant};
+
+use crate::{structures::block_info::BlockInfo, types::BlockInfoStream};
+
+pub async fn get_latest_block_info(
+    mut blockinfo_stream: BlockInfoStream,
```
Comment (grooviegermanikus) [inline core/src/utils.rs:10]:
```suggestion
    mut blockinfo_stream_of_commitment_level: BlockInfoStream,
```

File: core/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    commitment_config: CommitmentConfig,
+) -> BlockInfo {
+    let started = Instant::now();
+    loop {
+        match timeout(Duration::from_millis(500), blockinfo_stream.recv()).await {
+            Ok(Ok(block_info)) => {
+                if block_info.commitment_config == commitment_config {
```
Comment (grooviegermanikus) [inline core/src/utils.rs:17]: would you accept `block_info(finalized)` if you request `confirmed` or not?

File: core/src/structures/transaction_sent_info.rs [hunk 9]
```diff
@@ -7,7 +9,7 @@ pub type WireTransaction = Vec<u8>;
 pub struct SentTransactionInfo {
```
Comment (grooviegermanikus) [inline core/src/structures/transaction_sent_info.rs:9]: why is it called **Sent**? this reads to me like it happened already but the actually that struct contains the transaction to be sent

File: core/src/structures/transaction_sent_info.rs [hunk 10]
```diff
@@ -7,7 +9,7 @@ pub type WireTransaction = Vec<u8>;
...
     pub signature: Signature,
     pub slot: Slot,
-    pub transaction: WireTransaction,
+    pub transaction: Arc<WireTransaction>,
```
Comment (grooviegermanikus) [inline core/src/structures/transaction_sent_info.rs:12]: could that _Arc_ be avoided somehow? let me try for a minute

**Review by grooviegermanikus — CHANGES_REQUESTED** [review]
small readability comments

File: examples/custom-tpu-send-transactions/src/main.rs [hunk 11]
```diff
@@ -1,3 +1,519 @@
...
+    // start listning the cluster data and filling the cache
+    data_cache_service.listen(
+        endpoints.blocks_notifier.resubscribe(),
+        endpoints.blockinfo_notifier,
+        endpoints.slot_notifier.resubscribe(),
+        endpoints.cluster_info_notifier,
+        endpoints.vote_account_notifier,
+    );
+
+    let count = args.transaction_count.unwrap_or(10);
```
Comment (grooviegermanikus) [inline examples/custom-tpu-send-transactions/src/main.rs:347]: nit: should put that in Args struct

File: examples/custom-tpu-send-transactions/src/main.rs [hunk 12]
```diff
@@ -1,3 +1,519 @@
...
+            })
+            .collect_vec();
+        let tasks = transactions
+            .iter()
+            .map(|(signer, tx)| {
+                let rpc_client = rpc_client.clone();
+                let tx = tx.clone();
+                let signer = *signer;
+                tokio::spawn(
+                    async move { (signer, rpc_client.send_and_confirm_transaction(&tx).await) },
```
Comment (grooviegermanikus) [inline examples/custom-tpu-send-transactions/src/main.rs:126]: pls add .context

File: examples/custom-tpu-send-transactions/src/main.rs [hunk 13]
```diff
@@ -1,3 +1,519 @@
...
+
+    let tpu_service: TpuService = TpuService::new(
+        tpu_config,
+        validator_identity,
+        leader_schedule,
+        data_cache.clone(),
+    )
+    .await?;
+    let transaction_service_builder = TransactionServiceBuilder::new(
+        TxSender::new(data_cache.clone(), tpu_service.clone()),
```
Comment (grooviegermanikus) [inline examples/custom-tpu-send-transactions/src/main.rs:375]: the services are quite tangled. shouldn't it be easy to just wire up the TxSender? it requires recent blockhash and leader schedule

leader schedule could be:
1. full schedule from RPC
2. schedule for current epoch as text-file with each lead qualified by slots
3. hardcoded leaders (if the setup is just 1 test-validator or a small number of test nodes)

File: core/src/utils.rs (hunk 6 above)
Comment (godmodegalactus) [reply core/src/utils.rs:23]: This is just for debugging we do not need any additional log lines.

File: core/src/utils.rs (hunk 5 above)
Comment (godmodegalactus) [reply core/src/utils.rs:28]: The aim of this method is to wait till we get block of required commitment, this channel should never lag. If it lags then the logic is wrong.

File: services/src/tpu_utils/tpu_service.rs (hunk 2 above)
Comment (godmodegalactus) [reply services/src/tpu_utils/tpu_service.rs:131]: There is no method to collect as Vec. Specifying HashMap should be fine here.

**Pull request merged**